src/main/java/io/numaproj/pulsar/consumer/PulsarSource.java

```diff
@@ -0,0 +1,135 @@
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.List;
+
+@Slf4j
+@Component
+@ConditionalOnProperty(prefix = "spring.pulsar.consumer", name = "enabled", havingValue = "true")
+public class PulsarSource extends Sourcer {
+
+    // Mapping of readIndex to Pulsar messages that haven't yet been acknowledged
+    private final Map<Long, org.apache.pulsar.client.api.Message<byte[]>> messages = new ConcurrentHashMap<>();
+    private long readIndex = 0;
+    private Server server;
+
+    @Autowired
+    private PulsarClient pulsarClient;
+
+    // Consumer for Pulsar messages on hardcoded topic "my-topic"
+    private Consumer<byte[]> pulsarConsumer;
+
+    @PostConstruct // starts server automatically when the Spring context initializes
+    public void startServer() throws Exception {
+        // Create and start Pulsar consumer using the autowired PulsarClient.
+        pulsarConsumer = pulsarClient.newConsumer(Schema.BYTES)
+                .topic("my-topic")
+                .subscriptionName("my-topic-subscription")
+                .subscribe();
+
+        // Start the gRPC server for reading messages.
+        server = new Server(this);
+        server.start();
+        server.awaitTermination();
+    }
+
+    @Override
+    public void read(ReadRequest request, OutputObserver observer) {
+        long startTime = System.currentTimeMillis();
+
+        // minimal check: if outstanding messages haven't been acknowledged, do nothing.
+        if (!messages.isEmpty()) {
+            return;
+        }
+
+        for (int i = 0; i < request.getCount(); i++) {
+            // stop reading if timeout exceeds the allowed duration.
+            if (System.currentTimeMillis() - startTime > request.getTimeout().toMillis()) {
+                return;
+            }
+
+            try {
+                // Correctly cast the timeout to int as the receive method expects an int.
+                org.apache.pulsar.client.api.Message<byte[]> pMsg = pulsarConsumer
+                        .receive((int) request.getTimeout().toMillis(), TimeUnit.MILLISECONDS);
+                if (pMsg == null) {
+                    // no message received within timeout
+                    return;
+                }
+
+                // Log the consumed message (converting byte[] to String for logging clarity)
+                log.info("Consumed Pulsar message: {}", new String(pMsg.getValue()));
+
+                // Create a header map with additional Pulsar message id detail if needed.
+                Map<String, String> headers = new HashMap<>();
+                headers.put("pulsarMessageId", pMsg.getMessageId().toString());
```

Comment:
please investigate whether pulsar message object itself has headers that we need to pass to the next vertex.